Include microbenchmark for ES87TSDBDocValues

salvatore-campagna opened the pull request:
Here we include a microbenchmark for the encoding and decoding
of numeric values introduced by the new doc values format.
The benchmark runs for different values of the 'bitsPerValue'
parameter and exercises the encoding and decoding logic using
different blocks of numbers including both integer and floating
point values.

We also rollback a few changes meant to make the doc values
format work with different block sizes other than 128.
The `ForUtil` class is generated using a Python script and
we would like to keep the original algorithm that just works
for block size 128.

salvatore-campagna committed:
benchmark: include microbenchmark for ES87TSDBDocValues

Here we include a microbenchmark for the encoding and decoding
of numeric values introduced by the new doc values format.
The benchmark runs for different values of the 'bitsPerValue'
parameter and exercises the encoding and encoding logic
using different block of numbers including both integer
and floating point values.

We also rollback a few changes meant to make the doc values
format work with different block sizes other than 128.

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

salvatore-campagna committed:
fix: remove comment

jpountz commented on benchmarks/src/main/java/org/elasticsearch/benchmark/index/codec/tsdb/internal/DecodeDecreasingIntegerBenchmark.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public void setupIteration(int bitsPerValue) throws IOException {
+        this.input = generateMonotonicDecreasingInput(random.nextInt(1, 10), random.nextInt(1, 100));
```
Is it expected that it does not use `bitsPerValue`?

salvatore-campagna replied:
yes for decoding that is not used...it is required just for encoding.

jpountz commented on benchmarks/src/main/java/org/elasticsearch/benchmark/index/codec/tsdb/ES87TSDBDocValuesBenchmark.java:
```diff
@@ -0,0 +1,120 @@
+            "encode decreasing integer",
+            "encode floating point" }
+    )
+    private String mode;
```
JMH is going to test every combination of bitsPerValue and mode, but a non-zero number of bits per value and constant integers don't really make sense together? Maybe collapse these two benchmark params into a single one?

salvatore-campagna replied:
It is not clear to me what you means here:
* instead of using two parameters `bitsPerValue` and `mode` just use one (that encodes the combination) for all benchmarks?
*or just for the constant integer benchmark using the minimum number of bits necessary to represent the specific value of the constant integer?

I made two parameters and choose the verbose representation of the mode just to have a nicely formatted table of results at the end.

salvatore-campagna commented on benchmarks/src/main/java/org/elasticsearch/benchmark/index/codec/tsdb/internal/AbstractDocValuesForUtilBenchmark.java:
```diff
@@ -0,0 +1,64 @@
+        final long[] data = new long[blockSize];
+        data[blockSize - 1] = start;
+        for (int i = blockSize - 2; i >= 0; i--) {
+            data[i] = data[i + 1] - random.nextInt(step);
```
Yes correct...we are going backwards... 🤦

salvatore-campagna committed:
fix: decreasing instead of increasing

salvatore-campagna committed:
fix: use non-constant double values

salvatore-campagna committed:
fix: remove unused bitsPerValue parameter

salvatore-campagna committed:
fix: refactor random values generation logic

salvatore-campagna committed:
fix: use longs and include step 0

salvatore-campagna commented on benchmarks/src/main/java/org/elasticsearch/benchmark/index/codec/tsdb/internal/AbstractDocValuesForUtilBenchmark.java:
```diff
@@ -0,0 +1,64 @@
+
+public abstract class AbstractDocValuesForUtilBenchmark {
+    protected final DocValuesForUtil forUtil;
+    protected final Random random;
```
I guess you mean every benchmark instance needs to have its own Random instance right?

salvatore-campagna committed:
refactor: extract decoding and encoding logic

Extract the decoding and encoding logic in separate classes
reusing them for different benchmarks. Also separate constant
integer encoding and decoding and use suppliers to provide
data input.

salvatore-campagna committed:
fix: add missing classes

salvatore-campagna committed:
fix: add missing parenthesis

jpountz commented on benchmarks/src/main/java/org/elasticsearch/benchmark/index/codec/tsdb/internal/DecreasingIntegerSupplier.java:
```diff
@@ -0,0 +1,34 @@
+        for (int i = 0; i < size; i++) {
+            data[i] = random.nextLong(min, max);
+        }
+        return Arrays.stream(data).boxed().sorted(Collections.reverseOrder()).mapToLong(Long::longValue).toArray();
```
For monotonic sequences, I had imagined that the number of bits per value would be the max number of bits per value required to encode the delta between two consecutive values. But you're doing something different here, which I'm not sure makes much sense?

salvatore-campagna replied:
Yes I actually did that changing the min to be equal to zero and did't commit.

salvatore-campagna commented on benchmarks/src/main/java/org/elasticsearch/benchmark/index/codec/tsdb/ES87TSDBDocValuesConstantBenchmark.java:
```diff
@@ -0,0 +1,78 @@
+
+@Fork(value = 1)
+@Warmup(iterations = 1)
+@Measurement(iterations = 1)
```
Oh yes...I forgot to restore the right values here...I changed them to do a quick test run.

salvatore-campagna commented on benchmarks/src/main/java/org/elasticsearch/benchmark/index/codec/tsdb/internal/DecodeBenchmark.java:
```diff
@@ -0,0 +1,45 @@
+    @Override
+    public void setupInvocation(int bitsPerValue) {
+        this.dataInput = new ByteArrayDataInput(inputBuffer);
+        System.arraycopy(outputBuffer, 0, inputBuffer, 0, outputBuffer.length);
```
If I remember correctly the `reset` method still needs an array as an argument.

salvatore-campagna commented on benchmarks/src/main/java/org/elasticsearch/benchmark/index/codec/tsdb/internal/DecodeBenchmark.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Override
+    public void benchmark(int bitsPerValue) throws IOException {
+        forUtil.decode(bitsPerValue, this.dataInput, this.output);
```
Ok this is something I didn't know about JMH. If I understand correctly the idea is that if the output is not used the JVM might optimize away the code.

jpountz replied:
Correct.

salvatore-campagna committed:
fix: use 0 as minimum value

salvatore-campagna committed:
fix: consume the output using a BlackHole

salvatore-campagna committed:
fix: just reset the buffer in setupInvocation

salvatore-campagna committed:
fix: use correct values for warmup and iterations

salvatore-campagna committed:
fix: just pass the upper bound to 'nextLong'

jpountz reviewed (commented):
Thanks @salvatore-campagna this looks better now.

One thing that still looks awkward to me is having switch statements in methods that are annotated with `@Benchmark`: we would like these methods to only comprise the logic that we'd like to benchmark, without switch statements or virtual method calls that might add overhead?

salvatore-campagna commented:
> Thanks @salvatore-campagna this looks better now.
> 
> One thing that still looks awkward to me is having switch statements in methods that are annotated with `@Benchmark`: we would like these methods to only comprise the logic that we'd like to benchmark, without switch statements or virtual method calls that might add overhead?

It makes sense even if I thought after a few iterations the impact would be negligible.

salvatore-campagna committed:
refactor: remove switch from benchmask method

Also, include encoding and decoding benchmark for non sorted data.
We expect this to mimic fields storing gauge values.

salvatore-campagna committed:
chore: rename instance variable

salvatore-campagna committed:
Merge branch 'elastic:main' into feature/tsdb-compression-benchmark

salvatore-campagna commented:
@elasticsearchmachine run elasticsearch-ci/part-2 please.

salvatore-campagna commented:
@elasticsearchmachine run elasticsearch-ci/part-1 please.

salvatore-campagna committed:
fix: allocate output array at iteration time
